lib/Sema/TypeCheckType.cpp

```diff
@@ -5650,6 +5651,20 @@ TypeResolver::resolveCompositionType(CompositionTypeRepr *repr,
   return composition;
 }
 
+
+bool shouldDiagnoseImplicitlyUnwrappedOptionalType(TypeResolutionOptions options) {
+  if (
+      // to match `let _: (Int, any P!)`
+      options.is(TypeResolverContext::TupleElement) ||
+      // to match `let _: G<any P!>`
+      options.is(TypeResolverContext::ScalarGenericArgument) ||
+      // because  `let _ = (any P!) -> Void` is illegal, but not `let _ = any P!`
+      !options.is(TypeResolverContext::PatternBindingDecl)) {
+    return true;
+  }
+  return false;
+}
+
```

Comment:
i extracted the logic to the proposed function, this function now handles all the cases we discussed

i'm not happy about `!options.is(TypeResolverContext::PatternBindingDecl)`
but i needed something to prevent `let _ = any P!` from emitting: 

`using '!' is not allowed here; perhaps '?' was intended`

the the context in the case of `let _ = (any P!) -> Void` is None (surpassingly?) so i ended up doing it this way

Reply:
Please implement this function using a `switch` as in `resolveImplicitlyUnwrappedOptionalType`. This way an exhaustivity error will be raised if someone adds a new `TypeResolverContext` case, and the programmer will know to handle the case here.

Keep in mind that the idea is to *decouple* the "should diagnose?" logic from `resolveImplicitlyUnwrappedOptionalType` such that we can replace it with a call to this function — we are merely moving existing code.